element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpClientConnector.java

```diff
@@ -74,39 +83,52 @@ public TcpClientConnector(int numberOfThreads, int connectTimeoutMillis, int idl
 		workerGroup = new NioEventLoopGroup(numberOfThreads);
 		poolMap = new AbstractChannelPoolMap<SocketAddress, ChannelPool>() {
 
-			@Override protected ChannelPool newPool(SocketAddress key) {
+			@Override
+			protected ChannelPool newPool(SocketAddress key) {
 				Bootstrap bootstrap = new Bootstrap().group(workerGroup).channel(NioSocketChannel.class)
 						.option(ChannelOption.SO_KEEPALIVE, true).option(ChannelOption.AUTO_READ, true)
 						.option(ChannelOption.CONNECT_TIMEOUT_MILLIS, connectTimeoutMillis).remoteAddress(key);
 
-				// We multiplex over the same TCP connection, so don't acquire more than one connection per endpoint.
+				// We multiplex over the same TCP connection, so don't acquire
+				// more than one connection per endpoint.
 				// TODO: But perhaps we could make it a configurable property.
 				return new FixedChannelPool(bootstrap, new MyChannelPoolHandler(key), 1);
 			}
 		};
 	}
 
-	@Override public synchronized void stop() {
+	@Override
+	public synchronized void stop() {
 		if (null != workerGroup) {
 			workerGroup.shutdownGracefully(0, 1, TimeUnit.SECONDS).syncUninterruptibly();
 			workerGroup = null;
 		}
 	}
 
-	@Override public void destroy() {
+	@Override
+	public void destroy() {
 		stop();
 	}
 
-	@Override public void send(final RawData msg) {
+	@Override
+	public void send(final RawData msg) {
 		final ChannelPool channelPool = poolMap.get(new InetSocketAddress(msg.getAddress(), msg.getPort()));
 		Future<Channel> acquire = channelPool.acquire();
 		acquire.addListener(new GenericFutureListener<Future<Channel>>() {
 
-			@Override public void operationComplete(Future<Channel> future) throws Exception {
+			@Override
+			public void operationComplete(Future<Channel> future) throws Exception {
 				if (future.isSuccess()) {
 					Channel channel = future.getNow();
+
+					CorrelationContext context = NettyContextUtils.buildCorrelationContext(channel);
 					try {
 						channel.writeAndFlush(Unpooled.wrappedBuffer(msg.getBytes()));
+						if (null == context) {
+							NettyContextUtils.futureCorrelationContext(channel, msg);
```

Comment:
See comment above. With more netty knowledge we may change the order to first establis a TLS session and then send data, which would make this obsolete.